taichi/backends/vulkan/runtime.cpp

```diff
@@ -440,57 +447,88 @@ void VkRuntime::launch_kernel(KernelHandle handle, RuntimeContext *host_ctx) {
          /*export_sharing=*/false, AllocUsage::Storage});
   }
 
+  // Create context blitter
   auto ctx_blitter = HostDeviceContextBlitter::maybe_make(
       &ti_kernel->ti_kernel_attribs().ctx_attribs, host_ctx, device_,
       host_result_buffer_, ctx_buffer.get(), ctx_buffer_host.get());
 
   std::unordered_map<int, DeviceAllocation> ext_arrays;
+  std::unordered_map<int, size_t> ext_array_size;
 
+  // Prepare context buffers & arrays
+  const auto &args = ti_kernel->ti_kernel_attribs().ctx_attribs.args();
   if (ctx_blitter) {
     TI_ASSERT(ti_kernel->get_ctx_buffer_size());
 
-    const auto &args = ti_kernel->ti_kernel_attribs().ctx_attribs.args();
     int i = 0;
     for (auto &arg : args) {
       if (arg.is_array) {
-        if (arg.stride) {
+        if (host_ctx->is_device_allocation[i]) {
+          // NDArray
+          if (host_ctx->args[i]) {
+            ext_arrays[i] = *(DeviceAllocation *)(host_ctx->args[i]);
+          } else {
+            ext_arrays[i] = kDeviceNullAllocation;
+          }
```

Comment:
Do we need to update `ext_array_size` here?

Reply:
Nope, `ext_array_size` is only for ext arrs (for host device context blitter). For the runtime, with NDArrays, it does not need to know the buffer size. I will add a comment to clarify this.